src/compiler/checker.ts

```diff
@@ -8228,6 +8358,27 @@ module ts {
                             break;
                     }
                 }
+
+                if (node.type) {
+                    if (languageVersion >= ScriptTarget.ES6 && isSyntacticallyValidGenerator(node)) {
+                        let returnType = getTypeFromTypeNode(node.type);
+                        if (returnType === voidType) {
+                            error(node.type, Diagnostics.A_generator_cannot_have_a_void_type_annotation);

```

Comment:
I don't think people are likely to write an explicit `void` type annotation on a generator. Much more likely would be a primitive type like `string` or `number` or whatever the generator yields, but we aren't doing anything special for those.